File: 3849-equal-sum-grid-partition-i/equal-sum-grid-partition-i.py
class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:
        m = len(grid)
        n = len(grid[0])

        rowSum = [0] * m
        colSum = [0] * n

        total = 0
        for i in range(m):
            for j in range(n):
                total += grid[i][j]
                rowSum[i] += grid[i][j]
                colSum[j] += grid[i][j]

        if total % 2 != 0:
            return False

        upper = 0
        for i in range(m - 1):
            upper += rowSum[i]
            if upper == total - upper:
                return True

        left = 0
        for j in range(n - 1):
            left += colSum[j]
            if left == total - left:
                return True

        return False

File: 3849-equal-sum-grid-partition-i/equal-sum-grid-partition-i.cpp
class Solution {
public:
    typedef long long ll;
    bool canPartitionGrid(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();

        vector<ll> rowSum(m, 0);
        vector<ll> colSum(n, 0);

        ll total = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                total += grid[i][j];
                rowSum[i] += grid[i][j];
                colSum[j] += grid[i][j];
            }
        }
        if (total % 2 != 0) {
            return false;
        }
        ll upper = 0;
        for (int i = 0; i < m - 1; i++) {
            upper += rowSum[i];
            if (upper == total - upper) {
                return true;
            }
        }
        ll left = 0;
        for (int j = 0; j < n - 1; j++) {
            left += colSum[j];
            if (left == total - left) {
                return true;
            }
        }
        return false;
    }
};